File: Data Structures & Algorithms/daily-temperatures/submission-2.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        int[] result = new int[n];
        for (int i = 0;i<n ;i++) {
            while (!stack.isEmpty() && temperatures[stack.peek()] < temperatures[i]) {
                int prevTempDay = stack.pop();
                result[prevTempDay] = i - prevTempDay;
            }
            stack.push(i);
        }

        return result;
    }
}

File: Data Structures & Algorithms/minimum-stack/submission-2.kt
class MinStack() {
    val stack = ArrayDeque<Int>()
    val minStack = ArrayDeque<Int>()

    fun push(`val`: Int) {
         stack.addLast(`val`)
        if (minStack.isEmpty() || minStack.last() >= `val`){
            minStack.addLast(`val`)
        }
    }

    fun pop() {
        val last = stack.removeLast()
        if (last == minStack.last()) {
            minStack.removeLast()
        }
    }

    fun top(): Int {
        return stack.last()
    }

    fun getMin(): Int {
        return minStack.last()
    }
}
